<?xml version="1.0"?>
<!-- ROBOTIS Manipulator-X4 -->
<robot name="/gazebo/robotis/manipulator_x4" xmlns:xacro="http://www.ros.org/wiki/xacro">
  
  <!-- some parameters -->
  <xacro:property name="pi" value="3.141592654" />

  <!-- Import all Gazebo-customization elements, including Gazebo colors -->
  <xacro:include filename="$(find manipulator_x_description)/urdf/manipulator_x.gazebo" />
  <!-- Import Rviz colors -->
  <xacro:include filename="$(find manipulator_x_description)/urdf/materials.xacro" />

  <!-- World -->
  <link name="world">
  </link>

  <!-- World fixed joint-->
  <joint name="world_fixed" type="fixed">
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <parent link="world"/>
    <child link="link1"/>
  </joint>

  <!-- Base Plate -->
<!--
  <link name="base">
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://manipulator_x_description/meshes/manipulator_x_base.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>

    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://manipulator_x_description/meshes/manipulator_x_base.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>

    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <mass value="1.0"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0" iyy="1.0" iyz="0.0" izz="1.0" />
    </inertial>
  </link>
-->
  
  <!-- Base fixed joint -->
<!--
  <joint name="base_fixed" type="fixed">
    <parent link="base"/>
    <child link="link1"/>
  </joint>
-->
  <!-- Link 1 -->
  <link name="link1">
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://manipulator_x_description/meshes/manipulator_x_link1.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>

    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://manipulator_x_description/meshes/manipulator_x_link1.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>

    <inertial>
      <origin xyz="0.00025 0.00000 0.01780" rpy="0 0 0"/>
      <mass value="0.087"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0" />
      <inertia ixx="0.00001378" ixy="0.0" ixz="-0.00000025" iyy="0.00002381" iyz="0.0" izz="0.00002107" />
    </inertial>

  </link>

  <!-- Joint 1 -->
  <joint name="joint1" type="revolute">
    <parent link="link1"/>
    <child link="link2"/>
    <origin xyz="0.012 0.0 0.034" rpy="0 0 0"/>
    <axis xyz="0 0 1"/>
    <limit velocity="6.5" effort="1000" lower="${-pi*0.9}" upper="${pi*0.9}" />
    <dynamics damping="0.1"/>
  </joint>

  <transmission name="tran1">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint1">
      <hardwareInterface>PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor1">
      <hardwareInterface>PositionJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <!-- Link 2 -->
  <link name="link2">
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://manipulator_x_description/meshes/manipulator_x_link2.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>

    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://manipulator_x_description/meshes/manipulator_x_link2.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>

    <inertial>
      <origin xyz="0.00002 -0.00045 0.03293" rpy="0 0 0"/>
      <mass value="0.117"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0" />
      <inertia ixx="0.00004110" ixy="0.0" ixz="-0.00000025" iyy="0.00002381" iyz="0.0" izz="0.00002107" />
    </inertial>

  </link>

  <!-- Joint 2 -->
  <joint name="joint2" type="revolute">
    <parent link="link2"/>
    <child link="link3"/>
    <origin xyz="0.0 0.0 0.04235" rpy="0 0 0"/>
    <axis xyz="0 -1 0"/>
    <limit velocity="6.5" effort="1000" lower="${-pi*0.5}" upper="${pi*0.5}" />
    <dynamics damping="0.1"/>
  </joint>

  <transmission name="tran2">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint2">
      <hardwareInterface>PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor2">
      <hardwareInterface>PositionJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <!-- Link 3 -->
  <link name="link3">
    <collision>
      <origin xyz="0 -0.020 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://manipulator_x_description/meshes/manipulator_x_link3.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>

    <visual>
      <origin xyz="0 -0.020 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://manipulator_x_description/meshes/manipulator_x_link3.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>

    <inertial>
      <origin xyz="0.00966 -0.00034 0.11795" rpy="0 0 0"/>
      <mass value="0.156"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0" />
      <inertia ixx="0.00041624" ixy="0.00000017" ixz="-0.00004492" iyy="0.00042758" iyz="0.00000159" izz="0.00005317" />
    </inertial>

  </link>

  <!-- Joint 3 -->
  <joint name="joint3" type="revolute">
    <parent link="link3"/>
    <child link="link4"/>
    <origin xyz="0.024 0.0 0.12165" rpy="0 0 0"/>
    <axis xyz="0 -1 0"/>
    <limit velocity="6.5" effort="1000" lower="${-pi*0.5}" upper="${pi*0.5}" />
    <dynamics damping="0.1"/>
  </joint>

  <transmission name="tran3">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint3">
      <hardwareInterface>PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor3">
      <hardwareInterface>PositionJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <!-- Link 4 -->
  <link name="link4">
    <collision>
      <origin xyz="0 -0.020 0.0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://manipulator_x_description/meshes/manipulator_x_link4.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>

    <visual>
      <origin xyz="0 -0.020 0.0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://manipulator_x_description/meshes/manipulator_x_link4.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>

    <inertial>
      <origin xyz="0.11076 -0.00035 0.00050" rpy="0 0 0"/>
      <mass value="0.149"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0" />
      <inertia ixx="0.00003083" ixy="0.00000147" ixz="-0.00000222" iyy="0.00036703" iyz="-0.00000003" izz="0.00037667" />
    </inertial>

  </link>

  <!-- Joint 4 -->
  <joint name="joint4" type="revolute">
    <parent link="link4"/>
    <child link="link5"/>
    <origin xyz="0.1235 0.0 0.0" rpy="0 0 0"/>
    <axis xyz="0 -1 0"/>
    <limit velocity="6.5" effort="1000" lower="${-pi*0.5}" upper="${pi*0.5}" />
    <dynamics damping="0.1"/>
  </joint>

  <transmission name="tran4">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint4">
      <hardwareInterface>PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor4">
      <hardwareInterface>PositionJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <!-- Link 5 -->
  <link name="link5">
    <collision>
      <origin xyz="0 -0.020 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://manipulator_x_description/meshes/manipulator_x_link5.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>

    <visual>
      <origin xyz="0 -0.020 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://manipulator_x_description/meshes/manipulator_x_link5.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>

    <inertial>
      <origin xyz="0.03767 0.00262 -0.00043" rpy="0 0 0"/>
      <mass value="0.124"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0" />
      <inertia ixx="0.00004822" ixy="-0.00000148" ixz="0.00000024" iyy="0.00004342" iyz="0.00000014" izz="0.00005784" />
    </inertial>

  </link>

  <!-- Grip Joint -->
  <joint name="grip_joint" type="prismatic">
    <parent link="link5"/>
    <child link="grip_link"/>
    <origin xyz="0.064 0.0 0.0" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
    <limit velocity="6.5" effort="1000" lower="-0.01" upper="0.035" />
    <dynamics damping="0.1"/>
  </joint>

  <transmission name="tran7">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="grip_joint">
      <hardwareInterface>PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor7">
      <hardwareInterface>PositionJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <!-- Grip Link -->
  <link name="grip_link">
    <collision>
      <origin xyz="-0.064 -0.020 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://manipulator_x_description/meshes/manipulator_x_link_grip1.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>

    <visual>
      <origin xyz="-0.064 -0.020 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://manipulator_x_description/meshes/manipulator_x_link_grip1.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>

    <inertial>
      <origin xyz="0.02666 -0.00910 -0.02300" rpy="0 0 0"/>
      <mass value="0.06721"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0" />      
      <inertia ixx="0.00002875" ixy="-0.00000593" ixz="-0.00000006" iyy="0.00006646" iyz="-0.00000009" izz="0.00005474" />
    </inertial>

  </link>

  <!-- Grip Joint Sub -->
  <joint name="grip_joint_sub" type="prismatic">
    <parent link="link5"/>
    <child link="grip_link_sub"/>
    <origin xyz="0.064 0.0 0" rpy="0 0 0"/>
    <axis xyz="0 -1 0"/>
    <limit velocity="6.5" effort="1000" lower="-0.01" upper="0.035" />
    <dynamics damping="0.1"/>
  </joint>

  <transmission name="tran8">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="grip_joint_sub">
      <hardwareInterface>PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor8">
      <hardwareInterface>PositionJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <!-- Grip Link Sub -->
  <link name="grip_link_sub">
    <collision>
      <origin xyz="-0.064 -0.020 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://manipulator_x_description/meshes/manipulator_x_link_grip2.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>

    <visual>
      <origin xyz="-0.064 -0.020 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://manipulator_x_description/meshes/manipulator_x_link_grip2.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>

    <inertial>
      <origin xyz="0.02666 0.00910 -0.02300" rpy="0 0 0"/>
      <mass value="0.06721"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0" />      
      <inertia ixx="0.00002875" ixy="0.00000593" ixz="0.00000006" iyy="0.00006646" iyz="-0.00000009" izz="0.00005474" />
    </inertial>

  </link>
</robot>
